### PR TITLE
Show mev and extra data in block logs

### PR DESCRIPTION
**Pull request by benaadams**
image

## Changes

- Display the validator rewards 
- Display whether the block is mev
- Optionally display sanitised (for Utf8, no invalid, control chars etc) extra data (on by default)

#### What types of changes does your code introduce?

- [x] New feature (a non-breaking change that adds functionality)

#### Requires testing

- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs [hunk 1]
```diff
@@ -419,7 +419,21 @@ private void FireProcessingQueueEmpty()
...
+            bool isMev = false;
+            if (lastTx is not null && (lastTx.SenderAddress == beneficiary || _alternateMevPayees.Contains(lastTx.SenderAddress)))
+            {
+                // Mev reward with in last tx
+                beneficiary = lastTx.To;
+                isMev = true;
+            }
+            UInt256 beforeBalance = _stateReader.GetBalance(processingBranch.Root, beneficiary);
+            UInt256 afterBalance = _stateReader.GetBalance(suggestedBlock.StateRoot, beneficiary);
+            UInt256 rewards = beforeBalance < afterBalance ? afterBalance - beforeBalance : default;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs:435]: - Should all of this be in `BlockchainProcessor`, sounds like it should be more in the `UpdateStats`?
- As it is doing additional state reads, maybe it should be in background thread?

File: src/Nethermind/Nethermind.Merge.Plugin/IMergeConfig.cs [hunk 2]
```diff
@@ -69,4 +69,7 @@ The number of requests to the garbage collector (GC) to release the process memo
 
     [ConfigItem(Description = "[TECHNICAL] Simulate block production for every possible slot. Just for stress-testing purposes.", DefaultValue = "false", HiddenFromDocs = true)]
     bool SimulateBlockProduction { get; set; }
+
+    [ConfigItem(Description = "Whether to show the block's Extra Data field in logs", DefaultValue = "true")]
+    bool ShowExtraData { get; set; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/IMergeConfig.cs:74]: I think we don't need that as a param and just show it.

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs [hunk 3]
```diff
@@ -279,12 +285,118 @@ protected virtual bool IsNewHeadAlignedWithChain(Block newHeadBlock, ForkchoiceS
...
+
+        // Create the final string from the output buffer.
+        return outputPos > 0 ? new string(outputBuffer[..outputPos]) : string.Empty;
+    }
+
+    private static bool IsControlCharacter(Rune rune)
+    {
+        // Control characters are U+0000 to U+001F and U+007F to U+009F
+        return rune.Value <= 0x001F || (rune.Value >= 0x007F && rune.Value <= 0x009F);
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs:398]: Again sounds a bit out of place in `ForkChoiceUpdateHandler`

File: src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs]: Moved to UpdateStats background thread

File: src/Nethermind/Nethermind.Merge.Plugin/IMergeConfig.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Merge.Plugin/IMergeConfig.cs]: Removed as param

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Merge.Plugin/Handlers/ForkchoiceUpdatedHandler.cs]: Moved to Byte extensions; also moved ExtraData from ForkChoice to NewPayload

File: src/Nethermind/Nethermind.Config/BlocksConfig.cs [hunk 4]
```diff
@@ -53,5 +53,9 @@ public byte[] GetExtraDataBytes()
         {
             return _extraDataBytes;
         }
+
+        public string GasToken { get => GasTokenTicker; set => GasTokenTicker = value; }
+
+        public static string GasTokenTicker { get; set; } = "ETH";
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Config/BlocksConfig.cs:59]: EnergyWeb -> EWT
Exosama -> SAMA
JOC -> JOC
Volta - VT

File: src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs [hunk 5]
```diff
@@ -269,12 +274,19 @@ void IThreadPoolWorkItem.Execute()
...
             }
         }
+
+        // Help identify mev blocks when doesn't follow regular pattern
+        private static HashSet<AddressAsKey> _alternateMevPayees = new()
+        {
+            new Address("0xa83114A443dA1CecEFC50368531cACE9F37fCCcb"), // Extra data as: beaverbuild.org
+            new Address("0x9FC3da866e7DF3a1c57adE1a97c9f00a70f010c8"), // Extra data as: Titan (titanbuilder.xyz)
+            new Address("0x0b92619DdE55C0cbf828d32993a7fB004E00c84B"), // Extra data as: Builder+ www.btcs.com/builder
+        };
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Consensus/Processing/ProcessingStats.cs:290]: Do we want it hardcoded here? Aren't there more?

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 6]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
...
+                            outputBuffer[outputPos++] = ' ';
+                            shouldAddSpace = false;
+                        }
+
+                        int charsNeeded = rune.Utf16SequenceLength;
+                        if (outputPos + charsNeeded > outputBuffer.Length)
+                        {
+                            // Expand output buffer
+                            int newSize = outputBuffer.Length * 2;
+                            char[] newBuffer = new char[newSize];
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1182]: use pooled memory for new buffer

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 7]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
                     (BinaryPrimitives.ReverseEndianness(endIth), BinaryPrimitives.ReverseEndianness(ith));
             }
         }
+
+        public static string ToCleanUtf8String(this byte[] bytes)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1145]:
```suggestion
        public static string ToCleanUtf8String(this ReadOnlySpan<byte> bytes)
```

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 8]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
...
+
+                        rune.EncodeToUtf16(outputBuffer.Slice(outputPos));
+                        outputPos += charsNeeded;
+                        hasValidContent = true;
+                    }
+                    else
+                    {
+                        // Control character encountered; set flag to add space if needed
+                        if (hasValidContent)
+                            shouldAddSpace = true;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1195]:
```suggestion
                        shouldAddSpace |= hasValidContent;
```

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 9]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
...
+                    }
+                    index += bytesConsumed;
+                }
+                else if (status == OperationStatus.InvalidData)
+                {
+                    // Invalid data; set flag to add space if needed
+                    if (hasValidContent)
+                        shouldAddSpace = true;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1203]:
```suggestion
                    shouldAddSpace |= hasValidContent;
```

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 10]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
...
+                else if (status == OperationStatus.NeedMoreData)
+                {
+                    // Incomplete sequence at the end; break out of the loop
+                    break;
+                }
+                else
+                {
+                    // Unexpected status; treat as invalid data
+                    if (hasValidContent)
+                        shouldAddSpace = true;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1215]:
```suggestion
                    shouldAddSpace |= hasValidContent;
```

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs [hunk 11]
```diff
@@ -1140,5 +1141,91 @@ public static void ChangeEndianness8(Span<byte> bytes)
...
+                    }
+                    else
+                    {
+                        // Control character encountered; set flag to add space if needed
+                        if (hasValidContent)
+                            shouldAddSpace = true;
+                    }
+                    index += bytesConsumed;
+                }
+                else if (status == OperationStatus.InvalidData)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/Bytes.cs:1199]: This is same as `else`, so can be just removed?

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/NewPayloadHandler.cs [hunk 12]
```diff
@@ -451,6 +455,26 @@ private bool TryInsertDanglingBlock(Block block)
...
+        // and control chars to spaces and trim leading and trailing spaces.
+        string extraData = data.ToCleanUtf8String();
+
+        // If the cleaned text is less than half length of input size,
+        // output it as hex, else output the text.
+        return extraData.Length > data.Length / 2 ?
+            $"Extra Data: {extraData}" :
+            $"Hex: {data.ToHexString(withZeroX: true)}";
+    }
+
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Handlers/NewPayloadHandler.cs:477]: make it extension method?

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.Config/BlocksConfig.cs (hunk 4 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Config/BlocksConfig.cs:59]: Added

File: src/Nethermind/Nethermind.Merge.Plugin/Handlers/NewPayloadHandler.cs (hunk 12 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Merge.Plugin/Handlers/NewPayloadHandler.cs]: Done

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs (hunk 11 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Extensions/Bytes.cs]: Removed

File: src/Nethermind/Nethermind.Core/Extensions/Bytes.cs (hunk 6 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Extensions/Bytes.cs]: Done

**Pull request merged**